csrc/device_lower/lower2device.cpp

```diff
@@ -266,13 +266,17 @@ GpuLower::GpuLower(Fusion* fusion, const CompileParams& cparams)
            {"loadStoreOpInserter", loadStoreOpInserter},
            {"insertAllocations", insertAllocations},
            {"insertRawThreadSynchronization", insertRawThreadSynchronization},
+           // NOTE: the smem/register reuse pass occurs here
            {"reuseMemoryAllocations", reuseMemoryAllocations},
            {"insertWarThreadSynchronization", insertWarThreadSynchronization},
            {"DoubleBufferPass", DoubleBufferPass::run},
            {"rotateLoops", rotateLoops},
            {"UnrollPass", UnrollPass::runPass},
            {"processMisalignedVectorization", processMisalignedVectorization},
+           // NOTE: serial GridReduction introduced here. New syncs can be
+           // introduced here which could impact smem reuse
```

Comment:
We should probably factor out the logic for detecting serial grid reductions and insert those syncs much earlier than this point, so that they are available to `reuseMemoryAllocations`. This would let us re-use the wait sync instead of inserting a new one when we re-use prologue memory for epilogue.